File: codebase-query-contextualiser/function-calling/groq_service.py
import os
import json
import aiohttp
import asyncio
from typing import List, Dict, Optional
from dotenv import load_dotenv

class GroqService:
    _instance = None
    
    def __init__(self):
        # Load environment variables
        load_dotenv()
        self.api_key = os.getenv('GROQ_API_KEY')
        if not self.api_key:
            raise ValueError("GROQ_API_KEY not found in environment variables")
            
        self.base_url = 'https://api.groq.com/openai/v1/chat/completions'
        self.current_config = {
            "model": "mixtral-8x7b-32768",
            "temperature": 1,
            "max_tokens": 1024,
            "stream": False  # Changed to False for simpler response handling
        }

    @classmethod
    def get_instance(cls):
        if cls._instance is None:
            cls._instance = cls()
        return cls._instance

    def set_config(self, config: dict):
        self.current_config.update(config)

    def get_config(self):
        return self.current_config.copy()

    async def generate_response(self, messages: List[Dict]):
        try:
            headers = {
                'Content-Type': 'application/json',
                'Authorization': f'Bearer {self.api_key}'
            }
            
            payload = {
                **self.current_config,
                'messages': messages
            }

            async with aiohttp.ClientSession() as session:
                async with session.post(self.base_url, headers=headers, json=payload) as response:
                    if not response.ok:
                        error_text = await response.text()
                        raise Exception(f"HTTP error! status: {response.status}, body: {error_text}")

                    json_response = await response.json()
                    return json_response['choices'][0]['message']

        except Exception as e:
            print(f'Error in generate_response: {e}')
            raise

    def extract_json_from_response(self, result: str) -> str:
        try:
            start_idx = result.find('{')
            end_idx = result.rfind('}') + 1
            if start_idx >= 0 and end_idx > 0:
                json_str = result[start_idx:end_idx]
                json_str = (json_str
                    .replace('\\n', '')
                    .replace('\\"', '"')
                    .replace('\\', '')
                    .replace('"{', '{')
                    .replace('}"', '}'))
                return json_str
            return result
        except Exception as e:
            print(f'Error extracting JSON: {e}')
            return result

# Example usage
async def main():
    groq_service = GroqService.get_instance()
    messages = [
        {"role": "user", "content": "Hello, how are you?"}
    ]
    response = await groq_service.generate_response(messages)
    print(response)

if __name__ == "__main__":
    asyncio.run(main())

File: dev-ai/src/services/groqService.ts
import * as vscode from 'vscode';
import { GroqConfig, GroqMessage, GroqModel, GroqRequestBody } from '../types/groq-sdk';

export class GroqService {
    private static instance: GroqService;
    private apiKey: string;
    private baseUrl = 'https://api.groq.com/openai/v1/chat/completions';
    private currentConfig: GroqConfig = {
        model: "mixtral-8x7b-32768",
        temperature: 1,
        max_completion_tokens: 1024,
        top_p: 1,
        stream: true,
        stop: null
    };

    private constructor() {
        this.apiKey = process.env.GROQ_API_KEY || "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";
    }

    public static getInstance(): GroqService {
        if (!GroqService.instance) {
            GroqService.instance = new GroqService();
        }
        return GroqService.instance;
    }

    public setConfig(config: Partial<GroqConfig>) {
        this.currentConfig = { ...this.currentConfig, ...config };
    }

    public getConfig(): GroqConfig {
        return { ...this.currentConfig };
    }

    public async generateResponse(messages: GroqMessage[]): Promise<string> {
        try {
            const response = await fetch(this.baseUrl, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'Authorization': `Bearer ${this.apiKey}`
                },
                body: JSON.stringify({
                    model: this.currentConfig.model,
                    messages: messages,
                    temperature: this.currentConfig.temperature,
                    max_completion_tokens: this.currentConfig.max_completion_tokens,
                    stream: true
                })
            });

            if (!response.ok) {
                const errorText = await response.text();
                throw new Error(`HTTP error! status: ${response.status}, body: ${errorText}`);
            }

            const reader = response.body?.getReader();
            if (!reader) {
                throw new Error('Failed to get response reader');
            }

            const decoder = new TextDecoder();
            let result = '';
            let buffer = '';

            while (true) {
                const { value, done } = await reader.read();
                if (done) break;

                buffer += decoder.decode(value, { stream: true });
                const lines = buffer.split('\n');
                buffer = lines.pop() || '';

                for (const line of lines) {
                    if (line.trim() === '') continue;
                    if (line === 'data: [DONE]') continue;
                    if (!line.startsWith('data: ')) continue;

                    try {
                        const jsonStr = line.slice(5).trim();
                        if (!jsonStr) continue;
                        
                        const data = JSON.parse(jsonStr);
                        const content = data.choices?.[0]?.delta?.content;
                        if (content) {
                            result += content;
                        }
                    } catch (error: any) {
                        console.error('Error parsing JSON:', error.message);
                        console.error('Problematic line:', line);
                        continue;
                    }
                }
            }

            // Handle any remaining buffer
            if (buffer) {
                try {
                    if (buffer.startsWith('data: ') && buffer !== 'data: [DONE]') {
                        const jsonStr = buffer.slice(5).trim();
                        if (jsonStr) {
                            const data = JSON.parse(jsonStr);
                            const content = data.choices?.[0]?.delta?.content;
                            if (content) {
                                result += content;
                            }
                        }
                    }
                } catch (error: any) {
                    console.error('Error parsing remaining buffer:', error.message);
                }
            }

            return result;
        } catch (error: any) {
            console.error('Error in generateResponse:', error);
            throw new Error(`Failed to generate response: ${error.message}`);
        }
    }

    private extractJsonFromResponse(result: string): string {
        try {
            const startIdx = result.indexOf('{');
            const endIdx = result.lastIndexOf('}') + 1;
            if (startIdx >= 0 && endIdx > 0) {
                let jsonStr = result.slice(startIdx, endIdx)
                    .replace(/\\n/g, "")
                    .replace(/\\"/g, '"')
                    .replace(/\s+/g, " ")
                    .replace(/\\/g, "")
                    .replace(/"{/g, "{")
                    .replace(/}"/g, "}");
                return jsonStr;
            }
            return result;
        } catch (error) {
            console.error('Error extracting JSON:', error);
            return result;
        }
    }
} 